tests/junit.py

```diff
@@ -1,264 +1,128 @@
+        test_id: str = test[:4]
+        fail_str: str = ''
+        if test_id in ['t006', 't007']:
+            fail_str = 'No suitable backend:'
+        elif test_id in ['t008']:
+            fail_str = 'Available backend resources:'
+        elif test_id in ['t110', 't111', 't112', 't113', 't114']:
+            fail_str = 'Cannot grant CeedVector array access'
+        elif test_id in ['t115']:
+            fail_str = 'Cannot grant CeedVector read-only array access, the access lock is already in use'
+        elif test_id in ['t116']:
+            fail_str = 'Cannot destroy CeedVector, the writable access lock is in use'
+        elif test_id in ['t117']:
+            fail_str = 'Cannot restore CeedVector array access, access was not granted'
+        elif test_id in ['t118']:
+            fail_str = 'Cannot sync CeedVector, the access lock is already in use'
+        elif test_id in ['t215']:
+            fail_str = 'Cannot destroy CeedElemRestriction, a process has read access to the offset data'
+        elif test_id in ['t303']:
+            fail_str = 'Length of input/output vectors incompatible with basis dimensions'
+        elif test_id in ['t408']:
+            fail_str = 'CeedQFunctionContextGetData(): Cannot grant CeedQFunctionContext data access, a process has read access'
+        elif test_id in ['t409'] and contains_any(resource, ['memcheck']):
+            fail_str = 'Context data changed while accessed in read-only mode'
+
+        return fail_str, fail_str in stderr
+
+    def check_allowed_stdout(self, test: str) -> bool:
+        return test[:4] in ['t003']
 
-        # run tests
-        result = run(args.test, backends, args.mode)
 
-        # build output
-        if args.mode.lower() == "junit":
-            junit_batch = ''
-            try:
-                junit_batch = '-' + os.environ['JUNIT_BATCH']
-            except:
-                pass
-            output = Path('build') / (args.test + junit_batch + '.junit') if args.output is None else Path(args.output)
-
-            with output.open('w') as fd:
-                TestSuite.to_file(fd, [result])
-        elif args.mode.lower() != "tap":
-            raise Exception("output mode not recognized")
-
-        # check return code
-        for t in result.test_cases:
-            failures = len([c for c in result.test_cases if c.is_failure()])
-            errors = len([c for c in result.test_cases if c.is_error()])
-            if failures + errors > 0 and args.mode.lower() != "tap":
-                sys.exit(1)
+if __name__ == '__main__':
+    args = create_argparser().parse_args()
+
+    # run tests
+    mode: RunMode = RunMode(args.mode.lower())
+    result: TestSuite = run_tests(args.test, args.ceed_backends, mode, args.nproc, CeedSuiteSpec())
+
+    # build output
+    if mode is RunMode.JUNIT:
+        junit_batch: str = f'-{os.environ["JUNIT_BATCH"]}' if 'JUNIT_BATCH' in os.environ else ''
```

Comment:
This should also probably be an explicit argument? I forgot about this one when making the changes in Ratel